### PR TITLE
Make FirstOrNone, LastOrNone, SingleOrNone and ElementAtOrNone Lazy

### PR DESCRIPTION
**Pull request by FreeApophis**
`FirstOrNone` is an optimized version, otherwise we use the underlying Linq implementations

### CODE REVIEW

**Comment by FreeApophis** [comment]
Pull Request for Issue #13

File: Galaxus.Functional/Linq/OptionLinqExtensions.cs
Comment (SamuelBittmann) [inline Galaxus.Functional/Linq/OptionLinqExtensions.cs:63]: Couldn't we just write `source.Where(predicate).Select(Option<TSource>.Some).FirstOrDefault()` instead?

**Review by SamuelBittmann — CHANGES_REQUESTED** [review]
Hey @FreeApophis, thanks for your pull request. I assume your aim is to enable these extension methods to work with large or infinite collections (/generators). In order to document and verify this requirement I suggest to add corresponding unit tests for this. Other than this I really like the improvements :)

File: Galaxus.Functional/Linq/OptionLinqExtensions.cs
Comment (FreeApophis) [reply Galaxus.Functional/Linq/OptionLinqExtensions.cs:63]: You can, and it is actually the current implementation in Funcky [https://github.com/polyadic/funcky/blob/master/Funcky/Extensions/EnumerableExtensions/FirstOrNone.cs]. We tested the two versions against each other, and it looks like the optimization is quite a bit faster.

**Comment by FreeApophis** [comment]
It's not all about performance and infinite Generators but to stay true to the paradigm you are programming in. The current implementation is very procedural (if statments) and does not take advantage of the functional programming paradigm (stateless, lazy). 

I think it's great you are trying to introduce this style. A simple library like this is great testbed to learn the style of programming, and I want to give a little input from when you are little bit further down that road.

I'll adress your idea on monday with another commit.

**Comment by NoelWidmer** [comment]
What is the current state on this @phkiener and @SamuelBittmann? I can look into this once I am back from holidays (after the 19.10.20) if the PR is still open by then.

File: Galaxus.Functional/Linq/OptionLinqExtensions.cs
Comment (SamuelBittmann) [reply Galaxus.Functional/Linq/OptionLinqExtensions.cs:63]: I was able to reproduce your results with my own benchmark (see https://gist.github.com/SamuelBittmann/e9d2487f494e6f4a6db762fb2781c39f). At least when I was using `Enumerable.Range(...)` as the source. When I converted the range to a list first I got inverted results and the Linq implementation won (N is the number of elements in the enumerable):
image
For consistency reasons, and because we at Digitec Galaxus mostly use these methods together with lists, I would suggest we use the Linq version here. Or tell me if I did something wrong with the benchmark 😊

**Comment by SamuelBittmann** [comment]
> 
> 
> What is the current state on this @phkiener and @SamuelBittmann? I can look into this once I am back from holidays (after the 19.10.20) if the PR is still open by then.

I was gone for a few days, but I had time to look at the changes now. In my opinion this PR improves the existing code and I do not see any downsides. Therefore, I suggest we merge the PR as soon as all open discussions have been completed. @phkiener any inputs?

**Comment by SamuelBittmann** [comment]
@FreeApophis Thanks for the tests :)

**Comment by phkiener** [comment]
> [...] @phkiener any inputs?

Did not expect getting tagged twice! Is it my birthday? :D
The only thing I'd say is that

```csharp
something
    .Where(predicate)
    .Select(Option<T>.Some)
    .SomeOtherLinqOperationOrDefault()
```

takes a bit of brainpower to understand why it works - I am always a fan of being explicit. But if that is not an issue for anyone else here, I won't mind it. Other than that... go for it!

File: Galaxus.Functional/Linq/OptionLinqExtensions.cs
Comment (FreeApophis) [reply Galaxus.Functional/Linq/OptionLinqExtensions.cs:63]: Interesting test. Feel free to chose any implementation you see fit. The net gain is neglectable in most scnerios anyway.

**Comment by FreeApophis** [comment]
I just saw a few erronouse auto-formatting changes, I'll remove the line noise tomorrow. 

I'll also change the yield version to the Linq version.

@phkiener :) If you think thats a brainer...

Take a look at this: 
https://github.com/FreeApophis/RangeCalendarCSharp/blob/master/Calendar/ConsoleCalendar.cs

**Review by SamuelBittmann — APPROVED** [review]

**Comment by SamuelBittmann** [comment]
Alright, I'm happy with the current state. @FreeApophis Thanks for your efforts! @phkiener My review does not count since I don't have write access. Could you approve it if you are happy as well?

**Review by phkiener — APPROVED** [review]

**Comment by FreeApophis** [comment]
You need to merge it, I can't.

**Comment by phkiener** [comment]
Oh yeah, I totally forgot that merging also needs write-access... which is kinda obvious in hindsight, but oh well.
Anyways, it's now released as 3.2.0 ;)

**Pull request merged**
